RUM-6296 [Mobile Session Replay] Add documentation for fine grained masking overrides (#25849)
DataDog/documentation: content/en/real_user_monitoring/session_replay/mobile/privacy_options.md

By enabling Mobile Session Replay, you can automatically mask sensitive elements from being recorded through the RUM Mobile SDK. When data is masked, that data is not collected in its original form by Datadog's SDKs and thus is not sent to the backend.

> ## Fine-Grained Masking

Reviewer: ```suggestion
## Fine-grained masking
```

{{% /tab %}}
{{< /tabs >}}

> ## Legacy Masking - Deprecated

Reviewer: ```suggestion
```
Since this function no longer works, we should remove the section entirely.

{{% /tab %}}
{{< /tabs >}}

> ## Privacy Overrides

Reviewer: ```suggestion
## Privacy overrides
```

The sections above describe the global masking levels that apply to the entire application. However, it is also possible to override these settings at the view level. The same privacy levels as above are available for text and inputs, images, touches, and an additional setting to completely hide a specific view.

To ensure overrides are recognized properly, they should be applied as early as possible in the view lifecycle. This prevents scenarios where Session Replay might process a view before applying the overrides.

> Privacy overrides affect views and their descendants. This means that even if an override is applied to a view where it might have no immediate effect (e.g., applying an image override to a text input), the override still applies to all child views. 

Reviewer: ```suggestion
Privacy overrides affect views and their descendants. This means that even if an override is applied to a view where it might have no immediate effect (for example, applying an image override to a text input), the override still applies to all child views. 
```

> Overrides operate using a "nearest parent" principle: if a view has an override, it uses that setting; otherwise, it inherits the privacy level from the closest parent in the hierarchy with an override. If no parent has an override, the view defaults to the application's general masking level.

Reviewer: ```suggestion
Overrides operate using a "nearest parent" principle: if a view has an override, it uses that setting. Otherwise, it inherits the privacy level from the closest parent in the hierarchy with an override. If no parent has an override, the view defaults to the application's general masking level.
```

> ### Text and Input Override

Reviewer: ```suggestion
### Text and input override
```

To override text and input privacy, use `setSessionReplayTextAndInputPrivacy` on a view instance and pass a value from the `TextAndInputPrivacy` enum. Passing `null` removes the override.

{{< code-block lang="kotlin" filename="build.gradle" disable_copy="false" collapsible="true" >}}
    // Set an text and input override on your view
    myView.setSessionReplayTextAndInputPrivacy(TextAndInputPrivacy.MASK_SENSITIVE_INPUTS)
    // Remove an image override from your view
    myView.setSessionReplayTextAndInputPrivacy(null)
{{< /code-block >}}
{{% /tab %}}

{{% tab "iOS" %}}

To override text and input privacy, use `dd.sessionReplayOverrides.textAndInputPrivacy` on a view instance and set a value from the `TextAndInputPrivacyLevel` enum. Setting it to `nil` removes the override.

{{< code-block lang="swift" filename="AppDelegate.swift" disable_copy="false" collapsible="true" >}}
    // Set an text and input override on your view
    myView.dd.sessionReplayOverrides.textAndInputPrivacy = .maskSensitiveInputs
    // Remove an image override from your view
    myView.dd.sessionReplayOverrides.textAndInputPrivacy = nil
{{< /code-block >}}
{{% /tab %}}
{{< /tabs >}}

> ### Image Override

Reviewer: ```suggestion
### Image override
```

{{< tabs >}}
{{% tab "Android" %}}

To override image privacy, use `setSessionReplayImagePrivacy` on a view instance and pass a value from the `ImagePrivacy` enum. Passing `null` removes the override.

{{< code-block lang="kotlin" filename="build.gradle" disable_copy="false" collapsible="true" >}}
    // Set an image override on your view
    myView.setSessionReplayImagePrivacy(ImagePrivacy.MASK_ALL)
    // Remove an image override from your view
    myView.setSessionReplayImagePrivacy(null)
{{< /code-block >}}
{{% /tab %}}

{{% tab "iOS" %}}

To override image privacy, use `dd.sessionReplayOverrides.imagePrivacy` on a view instance and set a value from the `ImagePrivacyLevel` enum. Setting it to `nil` removes the override.

{{< code-block lang="swift" filename="AppDelegate.swift" disable_copy="false" collapsible="true" >}}
    // Set an image override on your view
    myView.dd.sessionReplayOverrides.imagePrivacy = .maskAll
    // Remove an image override from your view
    myView.dd.sessionReplayOverrides.imagePrivacy = nil
{{< /code-block >}}
{{% /tab %}}
{{< /tabs >}}

> ### Touch Override

Reviewer: ```suggestion
### Touch override
```

{{< tabs >}}
{{% tab "Android" %}}

To override touch privacy, use `setSessionReplayTouchPrivacy` on a view instance and pass a value from the `TouchPrivacy` enum. Passing `null` removes the override.

{{< code-block lang="kotlin" filename="build.gradle" disable_copy="false" collapsible="true" >}}
    // Set a touch override on your view
    view.setSessionReplayTouchPrivacy(TouchPrivacy.HIDE)
    // Remove a touch override from your view
    view.setSessionReplayTouchPrivacy(null)
{{< /code-block >}}
{{% /tab %}}

{{% tab "iOS" %}}

To override touch privacy, use `dd.sessionReplayOverrides.touchPrivacy` on a view instance and set a value from the `TouchPrivacyLevel` enum. Setting it to `nil` removes the override.

{{< code-block lang="swift" filename="AppDelegate.swift" disable_copy="false" collapsible="true" >}}
    // Set a touch override on your view
    myView.dd.sessionReplayOverrides.touchPrivacy = .hide
    // Remove a touch override from your view
    myView.dd.sessionReplayOverrides.touchPrivacy = nil
{{< /code-block >}}

{{% /tab %}}
{{< /tabs >}}

> ### Hidden Override

Reviewer: ```suggestion
### Hidden elements override
```

For sensitive elements that need to be completely hidden, use the `hidden` setting.

When an element is `hidden`, it is replaced by a placeholder labeled as "Hidden" in the replay, and its subviews are not recorded.

> It is important to note that marking a view as `hidden` will not prevent touch interactions from being recorded on that element. To hide touch interactions as well, use the [touch override][2] in addition to marking the element as hidden.

Reviewer: ```suggestion
**Note**: Marking a view as `hidden` does not prevent touch interactions from being recorded on that element. To hide touch interactions as well, use the [touch override](#touch-override) in addition to marking the element as hidden.
```

{{< tabs >}}
{{% tab "Android" %}}

Use `setSessionReplayHidden(hide = true)` to hide the element. Setting `hide` to `false` removes the override.

{{< code-block lang="kotlin" filename="build.gradle" disable_copy="false" collapsible="true" >}}
    // Mark a view as hidden
    myView.setSessionReplayHidden(hide = true)
    // Remove the override from the view
    myView.setSessionReplayHidden(hide = false)
{{< /code-block >}}

{{% /tab %}}

{{% tab "iOS" %}}
{{< code-block lang="swift" filename="AppDelegate.swift" disable_copy="false" collapsible="true" >}}
    // Mark a view as hidden
    myView.dd.sessionReplayOverrides.hide = true
    // Remove the override from the view
    myView.dd.sessionReplayOverrides.hide = false
{{< /code-block >}}

**Note**: Setting the `hidden` override to `nil` has the same effect as setting it to `false`—it disables the override.

{{% /tab %}}
{{< /tabs >}}

### Notes on WebViews

> • Privacy overrides, aside from the `hidden` and `touch` options, are not supported for WebViews. Primarily manage their privacy using the [browser SDK privacy settings][1].

Reviewer: ```suggestion
• Privacy overrides, aside from the `hidden` and `touch` options, are not supported for WebViews. You can primarily manage their privacy using the [browser SDK privacy settings][1].
```

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /real_user_monitoring/session_replay/privacy_options

> [2]: #touch-override

Reviewer: ```suggestion
```
